// oae-tickets-zendesk/lib/api.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var assert = require('assert');

var PublicationsAPI = require('oae-publications/lib/api');
var UserAPI = require('oae-principals/lib/api.user');
var Validator = require('oae-util/lib/validator').Validator;

var Config = require('./internal/config');
var Templates = require('./templates');
var Zendesk = require('oae-tickets-zendesk/lib/internal/zendesk');


/**
 * This module implements the public interface of the oae-tickets-zendesk
 * package.
 *
 * @param  {Object}  [options]  Optionally specify alternate dependency implementations.
 *
 * @api private
 */
var TicketsZendeskAPI = exports.TicketsZendeskAPI = function TicketsZendeskAPI(options) {
    // Ensure all public method calls receive a TicketsZendeskAPI instance as this.
    _.bindAll(this, 'createZendeskTicket');

    options = options || {};
    this.config = options.config || Config;
    this.zendesk = options.zendesk || Zendesk;
    this.userapi = options.userapi || UserAPI;
    this.publicationsAPI = options.publicationsAPI || PublicationsAPI;
};
_.extend(TicketsZendeskAPI.prototype, {

    'createZendeskTicket': function(ctx, oaeTicket, callback) {
        var client = this.zendesk.zendeskClientFromOAEConfig(ctx);
        if (client === undefined) {
            return callback({'err': 'Error creating zendesk client'});
        }

        var groupID = this.config.getZendeskGroupID(ctx.user().tenant.alias);
        if (groupID === undefined) {
            return callback({'err': 'No valid groupID is set'});
        }

        // Avoid getUser() callback function shadowing this
        var _this = this;

        // Need to fetch the user who created the ticket to send their details
        // with the ticket.
        var user = this.userapi.getUser(ctx, oaeTicket.createdBy, function(err, user) {

            if (err) {
                return callback({
                    'err': err,
                    'oaeTicket': oaeTicket,
                }, 'Error fetching ticket\'s user');
            }

            var publication = _this.publicationsAPI.getPublication(ctx, oaeTicket.publicationId, function(err, publication) {
                if (err) {
                    return callback({
                        'err': 'Error fetching ticket\'s publication',
                        'cause': err,
                        'oaeTicket': oaeTicket
                    });
                }

                var zendeskUser = _this._getOrCreateZendeskUser(client, user.displayName, user.email, function(err, zendeskUser) {

                    if (err) {
                        return callback({
                            'err': 'Error creating zendesk user',
                            'cause': err
                        })
                    }

                    var zendeskTicket = _this._zendeskTicket(oaeTicket, zendeskUser.id, groupID, user, ctx.user().tenant, publication);

                    client.tickets.create(zendeskTicket, function(err, req, result) {
                        if (err) {
                            return callback({
                                'err': 'Error creating ticket on zendesk.com',
                                'cause': err,
                                'oaeTicket': oaeTicket,
                                'zendeskTicket': zendeskTicket,
                                'publication': publication,
                                'client': client
                            });
                        }
                        return callback(null, result);
                    });
                });
            });
        });
    },

    /**
     * Get the zendesk user for the specified email address, creating it
     * if it doesn't already exit. The name is not used to find users,
     * but will be used when creating a user is necessary.
     *
     * @param  {Object}  client         A node-zendesk API client object
     * @param  {String}  name           The full name of the user
     * @param  {String}  email          The user's email
     * @param  {Object}  callback.err   Standard error object
     * @param  {Object}  callback.user  A Zendesk user object (no {user: {...}} wrapper)
     *
     * @api private
     */
    '_getOrCreateZendeskUser': function(client, name, email, callback) {
        var validator = new Validator();
        validator.check(email).isEmail();
        if (validator.hasErrors()) {
            return callback({
                err: 'Cannot search for an invalid email address',
                email: email
            });
        }

        // If query string is an email, zendesk seems to only match users w/
        // the exact matching email address.
        client.users.search({query: email}, function(err, status, data) {
            if (err) {
                return callback({
                    err: 'error searching for user',
                    cause: err,
                    email: email
                });
            }
            if (data.length) {
                // No need to create a user as one already exists
                return callback(null, data[0]);
            }

            // Create user
            client.users.create({
                    user: {
                        name: name,
                        email: email
                    }
                }, function(err, status, data) {
                    if (err) {
                        return callback({
                            err: 'error creating user',
                            cause: err,
                            name: name,
                            email: email
                        });
                    }
                    if (status !== 201) {
                        return callback({
                            err: 'Unexpected response from Zendesk API, expected 201 Created',
                            status: status,
                            data: data
                        });
                    }

                    return callback(null, data);
                }
            );
        });
    },

    '_zendeskTicket': function(oaeTicket, requesterID, groupID, user, tenant, publication) {
        return {
            ticket: {
                'group_id': groupID,
                'requester_id': requesterID,
                'external_id': oaeTicket.id,
                'description': Templates.zendeskTicketBodyTemplate({
                    'user': user,
                    'tenant': tenant,
                    'publication': publication,
                    'oaeTicket': oaeTicket,
                    'groupID': groupID
                })
            }
        };
    }
});

// Create the default instance of the api with deafult options.
var defaultTicketsZendeskAPI = new TicketsZendeskAPI();

// Export the public methods of the default zendesk ticket API for typical use

/**
 * Create a ticket on a tenant's Zendesk instance from an OAE ticket.
 *
 * The ticket is sent to the zendesk instance configured in the tenant
 * implied by the passed context. The ticket's reporter is the user who
 * created the OAE ticket. A zendesk user is created using the user's
 * email if a user with the email does not already exist in Zendesk.
 *
 * @param  {Context}        ctx                         The context of the current request
 * @param  {Ticket}         oaeTicket                   The ticket to forward to Zendesk
 * @param  {Function}       fn.callback                 Standard callback function
 * @param  {Object}         fn.callback.err             Error object containg the error code and the error message
 * @param  {Object}         fn.callback.zendeskTicket   The ticket created on zendesk.
 */
exports.createZendeskTicket = defaultTicketsZendeskAPI.createZendeskTicket;
