// oae-publications/lib/test/util.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');

var RestUtil = require('oae-rest/lib/util');

/**
 * Get all the publications for a user
 *
 * @param  {RestContext}    restContext             Standard REST Context object that contains the current tenant URL and the current user credentials
 * @param  {String}         userId                  The OAE user id for whom to retrieve the publications
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            Standard error object, if any
 * @param  {Publication[]}  callback.publications   An array of publications
 */
var getAllPublicationsForUser = module.exports.getAllPublicationsForUser = function(restContext, userId, callback, _publications, _nextToken) {
    _publications = _publications || [];

    getPublications(restContext, userId, _nextToken, 25, function(err, data) {
        if (err) {
            return callback(err);
        }

        _publications = _.union(_publications, data.results);
        if (data.nextToken) {
            return getAllPublicationsForUser(restContext, userId, callback, _publications, data.nextToken);
        } else {
            return callback(null, _publications);
        }
    });
};

/**
 * Get a publication by its ID
 *
 * @param  {RestContext}    restCtx                 Standard REST Context object that contains the current tenant URL and the current user credentials
 * @param  {String}         publicationId           The ID of the publication to retrieve
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            Standard error object, if any
 * @param  {Publication}    callback.publication    The retrieved publication
 */
var getPublication = module.exports.getPublication = function(restCtx, publicationId, callback) {
    RestUtil.RestRequest(restCtx, '/api/publications/' + RestUtil.encodeURIComponent(publicationId), 'GET', null, callback);
};

/**
 * Get the publications that are linked to a user of which he is an author
 *
 * @param  {RestContext}    restCtx                 Standard REST Context object that contains the current tenant URL and the current user credentials
 * @param  {String}         userId                  The ID of the user for which to retrieve the publications
 * @param  {String}         [start]                 The data of the publication from which to start returning this page of results
 * @param  {Number}         [limit]                 The number of publications that should be returned in this page
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            Standard error object, if any
 * @param  {Publication}    callback.publication    The publications for the given user
 */
var getPublications = module.exports.getPublications = function(restCtx, userId, start, limit, callback) {
    var params = {
        'start': start,
        'limit': limit
    };
    RestUtil.RestRequest(restCtx, '/api/publications/library/' + RestUtil.encodeURIComponent(userId), 'GET', params, callback);
};

/**
 * Get the publications that are linked to a user of which he is the creator
 *
 * @param  {RestContext}    restCtx                 Standard REST Context object that contains the current tenant URL and the current user credentials
 * @param  {String}         userId                  The ID of the user for which to retrieve the publications
 * @param  {String}         [start]                 The data of the publication from which to start returning this page of results
 * @param  {Number}         [limit]                 The number of publications that should be returned in this page
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            Standard error object, if any
 * @param  {Publication}    callback.publication    The publications for the given user
 */
var getPublicationsByCreator = module.exports.getPublicationsByCreator = function(restCtx, userId, start, limit, callback) {
    var params = {
        'start': start,
        'limit': limit
    };
    RestUtil.RestRequest(restCtx, '/api/publications/created/' + RestUtil.encodeURIComponent(userId), 'GET', params, callback);
};

/**
 * Create a publication
 *
 * @param  {RestContext}    restCtx                         Standard REST Context object that contains the current tenant URL and the current user credentials
 * @param  {String[]}       sourceIds                       A set of source identifiers that identifies this publication in third-party data sets
 * @param  {String}         displayName                     The title of this publication
 * @param  {String}         publicationType                 The type of this publication
 * @param  {Number}         date                            The date it was published
 * @param  {String[]}       authors                         The names of the people who co-authored this publication
 * @param  {Object}         [extra]                         A set of extra options
 * @param  {String}         [extra.thumbnailUri]            The thubmnail uri of the publication
 * @param  {String}         [extra.publisher]               The publisher of the publication
 * @param  {String}         [extra.openAccessType]          The open access type (e.g. green, gold)
 * @param  {String}         [extra.journalName]             The name of the journal the publication was published in
 * @param  {String}         [extra.issueNumber]             The issue number of the journal this publication was published in, or the revision of a book, ..
 * @param  {String}         [extra.pageBegin]               At which page this publication can be found in case it's published in a book or journal
 * @param  {String}         [extra.pageEnd]                 At which page the publication ends in case it's published in a book or journal
 * @param  {String}         [extra.funders]                 The funders
 * @param  {String}         [extra.department]              The department (e.g. Department of Zoology)
 * @param  {String}         [extra.contactEmail]            The email address that will be used to keep the creator of the publication updated
 * @param  {String}         [extra.useCambridgeAddendum]    Whether or not the Cambridge addendum is used
 * @param  {String}         [extra.comments]                Any comments on the publication
 * @param  {Object}         [file]                          The uploaded file
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    Standard error object, if any
 * @param  {Publication}    callback.data.result            Indicates if this publication was newly created, updated an existing one or was ignored
 * @param  {Publication}    callback.data.publication       The created publication
 */
var createPublication = module.exports.createPublication = function(restCtx, sourceIds, displayName, publicationType, date, authors, extra, file, callback) {
    extra = extra || {};
    var params = {
        'displayName': displayName,
        'publicationType': publicationType,
        'date': date,
        'authors': authors,
        'sourceIds': sourceIds,
        'thumbnailUri': extra.thumbnailUri,
        'publisher': extra.publisher,
        'openAccessType': extra.openAccessType,
        'journalName': extra.journalName,
        'issueNumber': extra.issueNumber,
        'pageBegin': extra.pageBegin,
        'pageEnd': extra.pageEnd,
        'funders': extra.funders,
        'department': extra.department,
        'contactEmail': extra.contactEmail,
        'useCambridgeAddendum': extra.useCambridgeAddendum,
        'comments': extra.comments,
        'acceptanceDate': extra.acceptanceDate,
        'file': file
    };
    RestUtil.RestRequest(restCtx, '/api/publications/create', 'POST', params, callback);
};

/**
 * Links a user with a publication
 *
 * @param  {RestContext}    restCtx             Standard REST Context object that contains the current tenant URL and the current user credentials
 * @param  {String}         publicationId       The ID of the publication to link
 * @param  {String}         authorName          The name of the author on the publication to replace with the user ID
 * @param  {String}         userId              The ID of the user that should be linked with this publication
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        Standard error object, if any
 */
var linkPublicationToUser = module.exports.linkPublicationToUser = function(restCtx, publicationId, authorName, userId, callback) {
    var params = {
        'authorName': authorName,
        'userId': userId
    };
    RestUtil.RestRequest(restCtx, '/api/publications/' + RestUtil.encodeURIComponent(publicationId) + '/link', 'POST', params, callback);
};

/**
 * Update a publication
 *
 * @param  {RestContext}    restCtx             Standard REST Context object that contains the current tenant URL and the current user credentials
 * @param  {String}         publicationId       The ID of the publication to link
 * @param  {Object}         publication         Object containing the updated publication
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        Standard error object, if any
 */
var updatePublication = module.exports.updatePublication = function(restCtx, publicationId, publication, callback) {
    var params = {'publication': publication};
    RestUtil.RestRequest(restCtx, '/api/publications/' + RestUtil.encodeURIComponent(publicationId), 'POST', params, callback);
};
