// oae-tickets-zendesk/lib/internal/config.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var util = require('util');

var ConfigAPI = require('oae-config');
var LogAPI = require('oae-logger');
var Validator = require('oae-util/lib/validator').Validator;

var TicketsZendeskConfig = ConfigAPI.config('oae-tickets-zendesk');

var ZENDESK_URL_FORMAT = 'https://%s.zendesk.com/api/v2/';

/**
 * This module is responsible for fetching and validating user supplied
 * configuration for the oae-tickets-zendesk package. All functions return
 * undefined if the user supplied configuration value is invalid.
 *
 * @api private
 */
var Config = exports.Config = function(options) {
    var that = {};

    options = options || {};
    var log = options.log || LogAPI.logger('oae-tickets-zendesk');
    var configAPI = options.configAPI || TicketsZendeskConfig;

    that.isZendeskForwardingEnabled = function(tenantAlias) {
        var isZendeskForwardingEnabled = that._getConfigValue(tenantAlias, 'forwardToZendesk');

        var validator = new Validator();
        validator.check().isBoolean(isZendeskForwardingEnabled);

        return that._stripInvalidConfigValue(isZendeskForwardingEnabled, 'forwardToZendesk', validator, tenantAlias);
    };

    that.getZendeskSubdomain = function(tenantAlias) {
        var subdomain = that._getConfigValue(tenantAlias, 'zendeskSubdomain');

        var validator = new Validator();
        validator.check(subdomain).regex('^[a-zA-Z0-9-]+$');

        return that._stripInvalidConfigValue(subdomain, 'zendeskSubdomain', validator, tenantAlias);
    }

    that.getZendeskEmail = function(tenantAlias) {
        var email = that._getConfigValue(tenantAlias, 'zendeskUserEmail');

        var validator = new Validator();
        validator.check(email).isEmail();

        return that._stripInvalidConfigValue(email, 'zendeskUserEmail', validator, tenantAlias);
    };

    that.getZendeskGroupID = function(tenantAlias) {
        var groupID = that._getConfigValue(tenantAlias, 'zendeskGroupID');

        var validator = new Validator();
        validator.check(groupID).isInt();

        return that._stripInvalidConfigValue(groupID, 'zendeskGroupID', validator, tenantAlias);
    };

    that.getZendeskApiToken = function(tenantAlias) {
        var apiToken = that._getConfigValue(tenantAlias, 'zendeskApiToken');

        var validator = new Validator();
        validator.check(apiToken).notEmpty();

        return that._stripInvalidConfigValue(apiToken, 'zendeskApiToken', validator, tenantAlias);
    };

    that.getZendeskUrl = function(tenantAlias) {
        var subdomain = that.getZendeskSubdomain(tenantAlias);

        if (subdomain === undefined) {
            return undefined;
        }

        return util.format(ZENDESK_URL_FORMAT, subdomain);
    };

    that._getConfigValue = function(tenantAlias, name) {
        return configAPI.getValue(tenantAlias, 'tickets-zendesk', name);
    };

    that._stripInvalidConfigValue = function(configValue, configKey, validator, tenantAlias) {
        if (validator.hasErrors()) {
            log().error({
                'err': validator.getErrors(),
                'tenantAlias': tenantAlias,
                'configKey': configKey,
                'configValue': configValue,
            }, util.format('"%s" config item was invalid', configKey));
            return undefined;
        }
        return configValue;
    };

    return that;
};

// Export the public methods of the default Config instance
var defaultConfig = new Config();

/**
 * Whether forwarding OAE tickets to Zendesk has been enabled.
 *
 * @return {Boolean|undefined}
 */
exports.isZendeskForwardingEnabled = defaultConfig.isZendeskForwardingEnabled;

/**
 * Gets the configured subdomain under zendesk.com that tickets should
 * be forwarded to.
 *
 * @return {String|undefined}   A subdomain
 */
exports.getZendeskSubdomain = defaultConfig.getZendeskSubdomain;

/**
 * Gets the configured email address of the user account that should be
 * used to access the zendesk API.
 *
 * @return {String|undefined}   An email address
 */
exports.getZendeskEmail = defaultConfig.getZendeskEmail;

/**
 * Gets the configured ID of the group that created tickets should belong
 * to in zendesk.
 *
 * @return {Number}     The integer group ID
 */
exports.getZendeskGroupID = defaultConfig.getZendeskGroupID;

/**
 * Gets the configured API token to use when authenticating with the
 * zendesk API.
 *
 * @return {String}     An API token.
 */
exports.getZendeskApiToken = defaultConfig.getZendeskApiToken;

/**
 * Gets the base URL of the zendesk API based on the configured zendesk
 * subdomain.
 *
 * @return {String}     The zendesk API's base url
 */
exports.getZendeskUrl = defaultConfig.getZendeskUrl;
